File: backend/app/models/topology.py
"""Topology models for Watchtower."""

from typing import Optional

from pydantic import BaseModel

from .connection import Connection, ExternalLink
from .device import Device


class Position(BaseModel):
    """Canvas position for a cluster."""

    x: float
    y: float


class Cluster(BaseModel):
    """A group of related devices."""

    id: str
    name: str
    cluster_type: str  # firewall, switch, server, network
    icon: str  # shield, switch, server, wifi
    position: Position
    device_ids: list[str] = []
    status: str = "active"  # active, planned


class Topology(BaseModel):
    """Complete network topology."""

    clusters: list[Cluster] = []
    devices: dict[str, Device] = {}
    connections: list[Connection] = []
    external_links: list[ExternalLink] = []

    # Aggregate stats
    total_devices: int = 0
    devices_up: int = 0
    devices_down: int = 0
    active_alerts: int = 0


class TopologySummary(BaseModel):
    """Quick topology stats for header display."""

    total_devices: int = 0
    devices_up: int = 0
    devices_down: int = 0
    devices_degraded: int = 0
    active_alerts: int = 0
    critical_alerts: int = 0
    warning_alerts: int = 0

File: frontend/src/types/topology.ts
import type { Connection, ExternalLink } from './connection'
import type { Device } from './device'

export interface Position {
  x: number
  y: number
}

export interface Cluster {
  id: string
  name: string
  cluster_type: string
  icon: string
  position: Position
  device_ids: string[]
  status: 'active' | 'planned'
}

export interface Topology {
  clusters: Cluster[]
  devices: Record<string, Device>
  connections: Connection[]
  external_links: ExternalLink[]
  total_devices: number
  devices_up: number
  devices_down: number
  active_alerts: number
}

export interface TopologySummary {
  total_devices: number
  devices_up: number
  devices_down: number
  devices_degraded: number
  active_alerts: number
  critical_alerts: number
  warning_alerts: number
}
